### ISSUE TITLE
Timeout for Device.open()

### ISSUE THREAD

**Issue by lampwins** [Juniper/py-junos-eznc#780]
I have been going through the code and scratching my head as to how to get `Device` to timeout in a reasonable amount of time when I first call `open()`. The specific scenario is to have netconf enabled on the device via ssh but the protect RE firewall filter does not allow netconf access. On such a device it takes upwards or 2 minutes to timeout the connection on `open()` which gets extremely annoying when I am trying to connect to many devices.

I would like to have the ability to pass a timeout value to `open()` to quickly throw an error on a connection that will never open in the first place.

**Comment by stacywsmith** [comment]
@lampwins 
We don't yet support passing an initial connection timeout argument, although I agree it would be beneficial.

As an alternative, you can set the `auto_probe` argument to the `Device()` instantiation or the `open()` method. The `auto_probe` argument attempts to make a TCP connection to the the `port` (default 830) TCP port. If the TCP connection can't be competed within the `auto_probe` value number of seconds, then a `jnpr.junos.exception.ProbeError` exception is raised.

Here's an example showing a `ProbeError` after 5 seconds because `r0` isn't really listening for NETCONF over SSH connections on port `666`.
```
>>> from jnpr.junos.device import Device
>>> dev = Device(host='r0', user='user', passwd='user123', port=666)
>>> dev.open(auto_probe=5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/media/sf_Automation/Projects/PyEZ/Repos/stacywsmith/py-junos-eznc/lib/jnpr/junos/device.py", line 1227, in open
    raise EzErrors.ProbeError(self)
jnpr.junos.exception.ProbeError: ProbeError(r0)
>>> 
```

**Comment by derfWayne** [comment]
John,
You can call dev.timeout and set it to the value you want to timeout at.

from jnpr.junos.device import Device
dev = Device(host='r0', user='user', passwd='user123', port=666)
dev.open(auto_probe=5)
dev.timeout = timeout

**Comment by lampwins** [comment]
@stacywsmith thank you. I have been playing with `auto_probe` and it does cover some of the use cases but I agree that a proper timeout on `open()` would be more appropriate. 

@derfWayne My understanding is that is the timeout for successive rpc's. My issue is the timeout of the initial device connection via `open()`.

**Comment by stacywsmith** [comment]
>My understanding is that is the timeout for successive rpc's. My issue is the timeout of the initial device connection via `open()`.

That's correct. The `dev.timeout` attribute sets the default timeout for RPC responses and doesn't effect the connection timeout (in spite of some incorrect documentation that says otherwise.)

**Comment by stacywsmith** [comment]
> I agree that a proper timeout on `open()` would be more appropriate.
Let's use this issue to track that.

**Comment by njayarama** [comment]
what is current timeout value of the open()? 
is there any plan to fix this issue, am hitting this one on my network. I need to increase the timeout value.

**Comment by attila123** [comment]
TL;DR: it should be possible to set ssh connect timeout via the "connecttimeout" parameter in ~/.ssh/config (or some other ssh config file with custom path).
I did some debugging and looks like:
- PyEZ uses the ncclient library for ssh connections. In ncclient SSHSession.connect() can make use of the "connecttimeout" parameter from an ssh config file, see https://github.com/ncclient/ncclient/blob/v0.6.6/ncclient/transport/ssh.py#L382 (for the current ncclient 0.6.6)
- According to PyEZ documentation https://junos-pyez.readthedocs.io/en/2.2.1/jnpr.junos.html#module-jnpr.junos.device the Device constructor takes an optional parameter:
"ssh_config (str) – OPTIONAL The path to the SSH configuration file. This can be used to load SSH information from a configuration file. By default ~/.ssh/config is queried."
I tested it by creating ~/.ssh/config with the following content:
```
connecttimeout=42
```
and re-run my debugging. The value 42 was read and applied in https://github.com/ncclient/ncclient/blob/v0.6.6/ncclient/transport/ssh.py#L403
BTW the ssh_config man page (see on your computer or e.g. at https://linux.die.net/man/5/ssh_config) says that the keywords are case-insensitive and indeed it is parsed like that (ncclient uses the config parsing of paramiko (ssh client library) which converts the keywords to lowercase).

**Comment by ipmonk** [comment]
This should be closed. 
we can pass conn_open_timeout variable to Device.open() 

        :param int conn_open_timeout:
            *OPTIONAL* To specify the timeout in seconds, which will
            be used while opening SSH connection to the device

if it times out, ncclient will raise SSHError("Could not open socket to %s:%s" % (host, port)), and pyez will raise jnpr.junos.exception.ConnectRefusedError as a result.

**Comment by dineshbaburam91** [comment]
conn_open_timeout argument supported
